### DISCUSSION TITLE
Question about memory usage and PSRAM support

### DISCUSSION THREAD

**Discussion by LostInCompilation** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#375] [Q&A] [answered]
I'm a bit confused why with DMA I can only use about 180k of memory.
When I have an ESP32-S3 with about 520k of SRAM, isn't it possible to use almost the full 520k with DMA? That would mean I could use many 64x64 panels (if slowly of course because of data transfer rate).
Where does the 180k memory limit come from?

Thanks

**Comment by mrcodetastic** [comment]
> I'm a bit confused why with DMA I can only use about 180k of memory. When I have an ESP32-S3 with about 520k of SRAM, isn't it possible to use almost the full 520k with DMA? 

Not really. ESP32 is like a PC with Windows. The Operating System (RTOS) and WiFi and BT stack use a chunk of that. Your sketch will also use memory. So whatever you have left is what you can use. There's no limit on how much memory can be used for DMA.... if you have the memory to spare.

I'm not sure where you got the 180kb number from, I guess from the Excel? It's just a rough number.

```
  ESP_LOGI("I2S-DMA", "Free heap: %d",   heap_caps_get_free_size(MALLOC_CAP_INTERNAL));
```

**Comment by LostInCompilation** [comment]
Yes I got the 180k from the spreadsheet. I didn't expect RtOS to use that much memory, but I will check with your posted function.

What's about external SPI RAM? Is external RAM supported for DMA and your lib? Then we would have way more capabilities to run dozens of panels.

**Comment by mrcodetastic** [comment, marked as answer]
> Yes I got the 180k from the spreadsheet. I didn't expect RtOS to use that much memory, but I will check with your posted function.

The 180k was a very rough figure based on some testing ages back. More than happy for people to challenge this and provide better numbers for the documentation.

> What's about external SPI RAM? Is external RAM supported for DMA and your lib? Then we would have way more capabilities to run dozens of panels.

PSRAM supported on ESP32-S3 currently. Must be an ESP32-S3 with OCTAL PSRAM in order to be 'fast' enough. Clockrate limited to 13Mhz, which isn't great for long chains, but isn't terrible either.

https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA#memory-constraints
